### PR TITLE
Implement atomic updates

### PR DESCRIPTION
**Pull request by maethu**
This is a followup of #80.

Issue with #80:

I'm unable to get a proper solrconfig.xml generated by c.r.solrinstance 5.x, which supports <updateLog/> for the atomic update feature.
So I went ahead using collective.receipe.solrinstance 6.0.0b3.

Issue with the current build:
To boost a document or index on add no longer works. 
I don't know really why, the request sent so solr contains the <doc boost="100".... But solr seems to ignore it. Any suggestions how to solve this?

**UPDATE:**

Since the AtomicUpdates and index time boosting does not work together (at least in solr 4.10.x), I implemented a new config option for the solr control panel atomic_updates. It's enabled by default.

Further I extended the docs with some important infos about the AtomicUpdate feature.

**UPDATE 2:**

Tests are now fine :laughing:

Update the PR description
Updated sphinx docu with AtomicUpdates
Added new control panel option atomic_updates.
Fixed tests according to the fact, that it's not possible to have index time boosting and atomic update enabled at the same time.

### CODE REVIEW

**Comment by maethu** [comment]
Yet another PR for the atomic update feature. 

The old PR https://github.com/collective/collective.solr/pull/82 was based on branch, which is already merged.

**Comment by tisto** [comment]
@maethu thanks! will try to find time to look into it in the next days...

**Comment by jone** [comment]
@maethu is this PR finished?
@lukasgraf could you maybe also take a look as this is a port of your code from ftw.solr?
Would be great to have this :stuck_out_tongue_winking_eye:

File: docs/CHANGES.rst [hunk 1]
```diff
@@ -4,6 +4,12 @@ Changelog
 4.1.1 (unreleased)
 ------------------
 
+- Ported atomic updates from ftw.solr.
+  This requires you need to update your solr config, load the new solr config and
```
Comment (lukasgraf) [inline docs/CHANGES.rst]: `This requires you need to update` -> `This requires you to update`

File: docs/features/atomic_updates.rst [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Partial indexing documents (AtomicUpdates)
+******************************************
+
+This means whenever possible, only the necessary / specified attributes get updated in Solr, and more importantly, re-indexed by Plone's indexers.
+
+With collective.receipe.solr a new configuration is introduced, called `updateLog`. ``updateLog`` is enabled by default and allows atomic updates. In detail it adds a new field ``_version_`` to the schema and also adds "<updateLog />" to your solr config.
```
Comment (lukasgraf) [inline docs/features/atomic_updates.rst]: `receipe` -> `recipe`

File: src/collective/solr/interfaces.py [hunk 3]
```diff
@@ -262,6 +262,20 @@ class ISolrSchema(Interface):
         required=False,
     )
 
+    atomic_updates = Bool(
+        title=_('label_atomic_update',
```
Comment (lukasgraf) [inline src/collective/solr/interfaces.py]: `label_atomic_update` -> `label_atomic_updates` (also in `.po[t]` files)

File: src/collective/solr/solr.py [hunk 4]
```diff
@@ -212,7 +214,31 @@ def deleteByQuery(self, query):
...
+        return self._schema
+
     def add(self, boost_values=None, **fields):
+        solr_config = queryUtility(ISolrConnectionConfig)
+        atomic_updates_enabled = getattr(solr_config, 'atomic_updates', True)
+
+        schema = self.get_schema()
+        uniqueKey = schema.get('uniqueKey', None)
+        if uniqueKey is None:
+            raise Exception("Could not get uniqueKey from Solr schema")
```
Comment (lukasgraf) [inline src/collective/solr/solr.py]: Plain `Exceptions` are a pain to deal with - we should probably introduce a common base class `SolrException` in `collective/solr/exceptions.py` (and make the existing `SolrInactiveException` inherit from that as well).

File: src/collective/solr/solr.py [hunk 5]
```diff
@@ -212,7 +214,31 @@ def deleteByQuery(self, query):
...
+
+        if uniqueKey not in fields.keys():
```
Comment (lukasgraf) [inline src/collective/solr/solr.py]: `not in fields.keys()` -> `not in fields`

File: src/collective/solr/solr.py [hunk 6]
```diff
@@ -225,11 +251,24 @@ def add(self, boost_values=None, **fields):
...
+            # Add update="set" attribute to each field except for the uniqueKey
+            # field.
+            if f == uniqueKey:
+                tmpl = '<field name="%s">%%s</field>' % self.escapeKey(f)
+                lst.append(tmpl % self.escapeVal(v))
+                continue
+
             if f in boost_values:
-                tmpl = '<field name="%s" boost="%s">%%s</field>' % (
+                tmpl = '<field name="%s" boost="%s" update="set">%%s</field>' % (
```
Comment (lukasgraf) [inline src/collective/solr/solr.py]: Hmm - since index time boosting doesn't work together with atomic updates, I think this can (should?) be handled differently here (no boosting if atomic updates enabled). But please double check, maybe also check with @buchi, I remember him having looked into that a bit more in depth.

File: src/collective/solr/tests/data/add_request_with_boost_values.txt [hunk 7]
```diff
@@ -1,7 +1,7 @@
 POST /solr/update HTTP/1.1
 Host: localhost
 Accept-Encoding: identity
-Content-Length: 124
+Content-Length: 125
 Content-Type: text/xml; charset=utf-8
 
-<add><doc boost="2"><field name="id" boost="0.5">500</field><field name="name" boost="5">python test doc</field></doc></add>
\ No newline at end of file
+<add><doc boost="2"><field name="id">500</field><field name="name" boost="5" update="set">python test doc</field></doc></add>
```
Comment (lukasgraf) [inline src/collective/solr/tests/data/add_request_with_boost_values.txt]: Same here - boosting + atomic updates?

File: src/collective/solr/tests/test_server.py [hunk 8]
```diff
@@ -210,13 +210,18 @@ def testPartialReindex(self):
         self.assertEqual(counts['review_state'], 2)
 
     def testReindexKeepsBoostValues(self):
+        # Disable atomic updates, on order to test the index time boosting.
```
Comment (lukasgraf) [inline src/collective/solr/tests/test_server.py]: `on order` -> `in order`

**Comment by lukasgraf** [comment]
@maethu other than the points above - looks good to me! :+1: 

Thanks for porting this :grin: :tada:

**Comment by tisto** [comment]
@maethu would you mind rebasing the pr and fixing the issues @lukasgraf brought up?

**Comment by maethu** [comment]
@tisto Gonna fix the issues today!

File: src/collective/solr/solr.py (hunk 4 above)
Comment (maethu) [inline src/collective/solr/solr.py]: imho this should not be part of this PR.

File: src/collective/solr/solr.py (hunk 6 above)
Comment (maethu) [inline src/collective/solr/solr.py]: The `update="set"` Attribut will be removed later, if atomic updates are enabled. And yes you're if this attribute is present the index-time-boosting feature is disabled.  --> https://github.com/collective/collective.solr/pull/86/files#diff-ce14634d17ed2770c78538cdc89f0fcdR268

**Comment by maethu** [comment]
@tisto It should be ready now!

**Comment by tisto** [comment]
Awesome! Will try to find time to look into it later today.

File: src/collective/solr/solr.py (hunk 4 above)
Comment (gforcada) [inline src/collective/solr/solr.py]: Here it is: https://github.com/collective/collective.solr/pull/103

**Comment by tisto** [comment]
@maethu can we make @gforcada's pr part of this pr and squash the commits into one?

**Comment by maethu** [comment]
@tisto You mean this one? https://github.com/collective/collective.solr/pull/103

**Comment by maethu** [comment]
@tisto I cherry-picked the commit from @gforcada and squashed all commits into one.

**Comment by lukasgraf** [comment]
I'm going to go ahead and merge this, since all those points now have been addressed.
